src/mscorlib/src/System/Reflection/IntrospectionExtensions.cs

```diff
@@ -2,28 +2,27 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-/*=============================================================================
-**
-**
-** 
-**
-**
-** Purpose: Get the underlying TypeInfo from a Type
-**
-**
-=============================================================================*/
+using System.Diagnostics;
 
 namespace System.Reflection
 {
     public static class IntrospectionExtensions
     {
         public static TypeInfo GetTypeInfo(this Type type)
         {
+            IReflectableType reflectableType = type as IReflectableType;
+            if (reflectableType != null)
+                return reflectableType.GetTypeInfo();
+
             if (type == null)
                 throw new ArgumentNullException(nameof(type));
 
-            var rcType = (IReflectableType)type;
-            return rcType.GetTypeInfo();
+            // This is bizarre but compatible with the desktop which casts "type" to IReflectableType without checking and
```

Comment:
I think that the CoreCLR implementation is much easier to follow, and it is also smaller - will be better candidate for inlining, etc.

(It is fine to add comment to it that the cast throws InvalidCastException.)